File: 02966. Divide Array Into Arrays With Max Difference/solution.js
/**
 * Problem: 2966. Divide Array Into Arrays With Max Difference
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 25 ms (Beats 100%)
 */

/**
 * Divides array into triplets with max difference <= k
 *
 * @param {number[]} nums - Array of numbers
 * @param {number} k - Max difference allowed per triplet
 *
 * @returns {number[][]} - Array of triplets or empty array if impossible
 */
const divideArray = (nums, k) => {
  // Create a sorted copy of the input array for processing
  const sortedNumbers = Int32Array.from(nums)

  // Sort the array in ascending order
  sortedNumbers.sort()

  // Initialize an empty array to hold the result triplets
  const result = []

  // Process array in groups of 3 elements
  for (let i = 0; i < sortedNumbers.length; i += 3) {
    // Check if current triplet exceeds maximum allowed difference
    if (sortedNumbers[i + 2] - sortedNumbers[i] > k) return []

    // Add valid triplet to result
    result.push([sortedNumbers[i], sortedNumbers[i + 1], sortedNumbers[i + 2]])
  }

  return result
}

File: 02966. Divide Array Into Arrays With Max Difference/solution.go
/**
 * Problem: 2966. Divide Array Into Arrays With Max Difference
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 6 ms (Beats 100%)
 */

func divideArray(nums []int, k int) [][]int {
	// Find minimum and maximum values in the array
	minVal, maxVal := nums[0], nums[0]
	for _, val := range nums[1:] {
		if val < minVal {
			minVal = val
		}
		if val > maxVal {
			maxVal = val
		}
	}
	
	// Count frequency of each number using counting sort
	counts := make([]uint32, maxVal-minVal+1)
	for _, val := range nums {
		counts[val-minVal]++
	}
	
	// Reconstruct sorted array from counts
	index := 0
	for offset, count := range counts {
		for range count {
			nums[index] = offset + minVal
			index++
		}
	}
	
	// Create result arrays of size 3
	result := make([][]int, len(nums)/3)
	index-- // Adjust index from previous loop where index == len(nums)
	
	// Check if triplets satisfy max difference constraint (working backwards)
	for index > 0 && nums[index]-nums[index-2] <= k {
		result[index/3] = nums[index-2 : index+1]
		index -= 3
	}
	
	// Return nil if not all numbers were processed (constraint violated)
	if index > 0 {
		return nil
	}

	return result
}
